### PR TITLE
Fixed Homunculus Amistr skills

### PR DESCRIPTION
**Pull request by Daegaladh**
* **Addressed Issue(s)**: N/A

* **Server Mode**: Both

* **Description of Pull Request**: Fixes Amistr skills behavior:

    * HAMI_CASTLE (Castling):
      * Removed animation from the master (on official servers, the animations is shown only on the homunculus)
      * Fixes changing all monsters targeting the master to target the homunculus instead (on official servers, it only changes the target on ONE monster at a time)

    * HAMI_DEFENCE (Amistr Bulwark):
      * Fixed the effect not being removed from the
      * Fixed effect of SC_DEFENCE status (on official servers increases the VIT and as result VIT DEF on the master, and plain DEF on the homunculus, on both Renewal and Pre-renewal)
      * Added missing flags for SC_DEFENCE status:
          * Pre-renewal: Not getting removed from the homunculus on Teleport, Map change and Vaporize
          * Renewal: Not getting removed from the homunculus on Map change and Vaporize
          * Both: Not getting removed from both the homunculus and the master on Relog

    * HAMI_BLOODLUST (Bloodlust):
      * Fixed healing effect showing (on official servers, the hp is added to the health bar, but no healing effect is shown)
      * Fixed leeching chance (on official servers is actually 9 * level on both Renewal and Pre-renewal, despite the in-game description)
      * Added missing flags for SC_BLOODLUST status:
          * Pre-renewal: Not getting removed from the homunculus on Teleport, Map change and Vaporize
          * Renewal: Not getting removed from the homunculus on Map change and Vaporize
          * Both: Not getting removed from both the homunculus and the master on Relog

### CODE REVIEW

File: src/map/unit.cpp [hunk 1]
```diff
@@ -3144,20 +3144,29 @@ int unit_changetarget(struct block_list *bl, va_list ap) {
...
 	//unit_attack(bl, target->id, ud->state.attack_continue);
 	return 0;
 }
 
+/**
+ * Changes the target of a unit
+ * @param ud: Unit data
+ * @param target_id: New target ID
+ **/
+void unit_change_target(unit_data& ud, int target_id) {
```
Comment (aleos89) [inline src/map/unit.cpp:3158]: Should we check for invalid target_id? Just something along the lines of `> 0` or something?

File: src/map/unit.cpp [hunk 2]
```diff
@@ -3144,20 +3144,29 @@ int unit_changetarget(struct block_list *bl, va_list ap) {
...
+	if (ud.bl->type == BL_MOB)
+		reinterpret_cast<mob_data*>(ud.bl)->target_id = target_id;
+	if (ud.target_to)
+		ud.target_to = target_id;
+	else
+		ud.target_to = 0;
```
Comment (aleos89) [inline src/map/unit.cpp:3164]:
```suggestion
```
This won't be needed if you check target_id's value.

**Review by aleos89 — APPROVED** [review]

**Comment by AoShinRO** [comment]
finally anyone remember homunculus exists haha

File: src/map/unit.hpp [hunk 3]
```diff
@@ -172,8 +172,8 @@ void unit_free_pc(map_session_data *sd);
 #define unit_remove_map(bl,clrtype) unit_remove_map_(bl,clrtype,__FILE__,__LINE__,__func__)
 int unit_remove_map_(struct block_list *bl, clr_type clrtype, const char* file, int line, const char* func);
 int unit_free(struct block_list *bl, clr_type clrtype);
-int unit_changeviewsize(struct block_list *bl,short size);
 int unit_changetarget(struct block_list *bl,va_list ap);
+void unit_change_target(unit_data& ud, int target_id);
```
[the remark this answers is not in the document]
Comment (AoShinRO) [reply src/map/unit.hpp:176]: int unit_changetarget(struct block_list *bl,va_list ap); i think it's a subfunction from  map_foreach... maybe change  it  to  int unit_changetarget_sub(struct block_list *bl,va_list ap);

File: src/map/battle.cpp [hunk 4]
```diff
@@ -1900,8 +1900,8 @@ int64 battle_calc_damage(struct block_list *src,struct block_list *bl,struct Dam
 			}
 		}
 
-		if ((sce = sc->getSCE(SC_BLOODLUST)) && flag&BF_WEAPON && damage > 0 && rnd()%100 < sce->val3)
-			status_heal(src, damage * sce->val4 / 100, 0, 3);
+		if ((sce = sc->getSCE(SC_BLOODLUST)) && flag & BF_WEAPON && damage > 0 && rnd() % 100 < sce->val3)
```
Comment (Atemo) [inline src/map/battle.cpp:1903]:
```suggestion
		if ((sce = sc->getSCE(SC_BLOODLUST)) && flag & BF_WEAPON && damage > 0 && rnd_chance( sce->val3, 100 ))
```

File: src/map/status.cpp [hunk 5]
```diff
@@ -11489,16 +11486,21 @@ int status_change_start(struct block_list* src, struct block_list* bl,enum sc_ty
...
 #endif
 			break;
 		case SC_BLOODLUST:
-			val2 = 20+10*val1; // Atk rate change.
-			val3 = 3*val1; // Leech chance
-			val4 = 20; // Leech percent
+			// Atk rate change
+			val2 = 20 + (10 * val1);
+			// Leech chance
+			// It's actually 9 * level on both modes, despite the description
```
Comment (Atemo) [inline src/map/status.cpp:11500]: What do the “both modes” refer to?

File: src/map/unit.cpp [hunk 6]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
 
+/**
+ * Changes the target of a unit
+ * @param ud: Unit data
+ * @param target_id: New target ID
+ **/
+void unit_change_target(unit_data& ud, int target_id) {
+	if (target_id <= 0)
+		return;
+
```
Comment (AoShinRO) [inline src/map/unit.cpp:3163]:
```suggestion
        if(status_isdead(map_id2bl(target_id))
             return;
```

File: src/map/status.cpp (hunk 5 above)
Comment (Daegaladh) [reply src/map/status.cpp:11500]: pre-re and renewal

File: src/map/unit.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Daegaladh) [reply src/map/unit.hpp:176]: I used the naming as in unit_stopattack and unit_stop_attack, but maybe you're right

File: src/map/skill.cpp [hunk 7]
```diff
@@ -8239,8 +8239,10 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 		break;
 	case HLIF_AVOID:
 	case HAMI_DEFENCE:
-		sc_start(src,bl,type,100,skill_lv,skill_get_time(skill_id,skill_lv)); // Master
-		clif_skill_nodamage(src,src,skill_id,skill_lv,sc_start(src,src,type,100,skill_lv,skill_get_time(skill_id,skill_lv))); // Homunc
+		// Master
+		sc_start(src, bl, type, 100, skill_lv, skill_get_time(skill_id, skill_lv));
+		// Homunculus
+		clif_skill_nodamage(src, src, skill_id, skill_lv, sc_start(src, src, type, 100, skill_lv, skill_get_time(skill_id, skill_lv)));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:8245]: No change except comment?

File: src/map/skill.cpp [hunk 8]
```diff
@@ -10673,23 +10675,28 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
 	case HAMI_CASTLE:	//[orn]
 		if (src != bl && rnd_chance(20 * skill_lv, 100)) {
-			int x = src->x, y = src->y;
-
-			// Move source
-			if (unit_movepos(src,bl->x,bl->y,0,0)) {
-				clif_skill_nodamage(src,src,skill_id,skill_lv,1); // Homunc
+			// Get one of the monsters targeting the player and set the homunculus as its new target
+			block_list* tbl = battle_getenemy(bl, BL_MOB, AREA_SIZE);
+			if (tbl != nullptr) {
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10680]:
```suggestion
			if (block_list* tbl = battle_getenemy(bl, BL_MOB, AREA_SIZE); tbl != nullptr) {
```

File: src/map/skill.cpp [hunk 9]
```diff
@@ -10673,23 +10675,28 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
+				unit_data *ud = unit_bl2ud(tbl);
+				if (ud != nullptr)
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10682]:
```suggestion
				if (unit_data *ud = unit_bl2ud(tbl); ud != nullptr)
```

File: src/map/skill.cpp [hunk 10]
```diff
@@ -10673,23 +10675,28 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
+					unit_change_target(*ud, src->id);
+			}
+
+			int16 x = src->x, y = src->y;
+			// Move homunculus
+			if (unit_movepos(src, bl->x, bl->y, 0, 0)) {
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10688]:
```suggestion
			if (unit_movepos(src, bl->x, bl->y, 0, false)) {
```

File: src/map/unit.hpp [hunk 11]
```diff
@@ -172,8 +172,8 @@ void unit_free_pc(map_session_data *sd);
 #define unit_remove_map(bl,clrtype) unit_remove_map_(bl,clrtype,__FILE__,__LINE__,__func__)
 int unit_remove_map_(struct block_list *bl, clr_type clrtype, const char* file, int line, const char* func);
 int unit_free(struct block_list *bl, clr_type clrtype);
-int unit_changeviewsize(struct block_list *bl,short size);
-int unit_changetarget(struct block_list *bl,va_list ap);
+int unit_changetarget(block_list *bl,va_list ap);
+void unit_change_target(unit_data& ud, int target_id);
```
Comment (Lemongrass3110) [inline src/map/unit.hpp:176]:
```suggestion
void unit_changetarget(unit_data& ud, block_list& target);
```
We use C++ now, function overloading is fine.

File: src/map/unit.cpp [hunk 12]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
 	//unit_attack(bl, target->id, ud->state.attack_continue);
 	return 0;
 }
 
+/**
+ * Changes the target of a unit
+ * @param ud: Unit data
+ * @param target_id: New target ID
+ **/
+void unit_change_target(unit_data& ud, int target_id) {
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3160]:
```suggestion
void unit_changetarget(unit_data& ud, block_list& target) {
```

File: src/map/unit.cpp [hunk 13]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
-	if (bl->type == BL_MOB)
-		(BL_CAST(BL_MOB,bl))->target_id = target->id;
-	if (ud->target_to)
-		ud->target_to = target->id;
-	else
-		ud->target_to = 0;
-	if (ud->skilltarget)
-		ud->skilltarget = target->id;
-	unit_set_target(ud, target->id);
+	unit_change_target(*ud, target->id);
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3149]:
```suggestion
	unit_changetarget(*ud, *target);
```

File: src/map/skill.cpp [hunk 14]
```diff
@@ -10673,23 +10675,28 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
...
-
-			// Move source
-			if (unit_movepos(src,bl->x,bl->y,0,0)) {
-				clif_skill_nodamage(src,src,skill_id,skill_lv,1); // Homunc
+			// Get one of the monsters targeting the player and set the homunculus as its new target
+			block_list* tbl = battle_getenemy(bl, BL_MOB, AREA_SIZE);
+			if (tbl != nullptr) {
+				unit_data *ud = unit_bl2ud(tbl);
+				if (ud != nullptr)
+					unit_change_target(*ud, src->id);
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:10683]:
```suggestion
					unit_changetarget(*ud, *src);
```

File: src/map/unit.cpp [hunk 15]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
 }
 
+/**
+ * Changes the target of a unit
+ * @param ud: Unit data
+ * @param target_id: New target ID
+ **/
+void unit_change_target(unit_data& ud, int target_id) {
+	if (target_id <= 0 || status_isdead(map_id2bl(target_id)))
+		return;
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3162]:
```suggestion
	if (status_isdead(target))
		return;
```
You need to merge the master branch and then apply this suggestion.

File: src/map/unit.cpp [hunk 16]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
+
+	if (ud.bl->type == BL_MOB)
+		reinterpret_cast<mob_data*>(ud.bl)->target_id = target_id;
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3165]:
```suggestion
		reinterpret_cast<mob_data*>(ud.bl)->target_id = target->id;
```

File: src/map/unit.cpp [hunk 17]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
+	if (ud.target_to > 0)
+		ud.target_to = target_id;
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3167]:
```suggestion
		ud.target_to = target->id;
```

File: src/map/unit.cpp [hunk 18]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
+	if (ud.skilltarget > 0)
+		ud.skilltarget = target_id;
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3169]:
```suggestion
		ud.skilltarget = target->id;
```

File: src/map/unit.cpp [hunk 19]
```diff
@@ -3132,32 +3132,44 @@ int unit_counttargeted(struct block_list* bl)
...
+	unit_set_target(&ud, target_id);
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:3170]:
```suggestion
	unit_set_target(&ud, target->id);
```

**Review by Lemongrass3110 — APPROVED** [review]

File: src/map/skill.cpp (hunk 7 above)
Comment (Daegaladh) [reply src/map/skill.cpp:8245]: Yeah, just spacing and the comments

**Pull request merged**